lib/backend/k8s/syncer.go

```diff
@@ -280,18 +291,34 @@ func (syn *kubeSyncer) readFromKubernetesAPI() {
 	var event watch.Event
 	var kvp *model.KVPair
 	var opts metav1.ListOptions
-	var openWatchers []watch.Interface
+	var openWatchers = map[string]watch.Interface{}
 	closeWatchers := func() {
 		for _, w := range openWatchers {
 			log.WithField("watcher", w).Debug("Closing old watcher.")
 			w.Stop()
 		}
-		openWatchers = nil
+		openWatchers = map[string]watch.Interface{}
 	}
 
 	// Always perform an initial snapshot.
 	needsResync := true
 
+	watchEventContinue := func(event watch.Event, key string) bool {
+		if event.Type == watch.Error {
+			log.Warnf("Event requires resync: %+v", event)
+			needsResync = true
+			return true
+		} else if event.Object == nil {
+			log.Warnf("Need new %v watch: %+v", key, event)
+			w := openWatchers[key]
+			log.WithField("watcher", w).Debug("Closing old watcher.")
+			w.Stop()
```

Comment:
Ah, all watchers are closed whenever we resync.  I think my general feeling of unease with this bit of code is because there's a lot of blurring of responsibilities / side-effects here.  For example, this function: sets a flag defined outside of its inner scope, returns a different flag, and sometimes closes a watcher but not always.

I'd rather have clearly defined lines in terms of responsibility: 

```
// Returns whether this event triggers a full resync.
func eventNeedsResync(e watch.Event) bool

// Returns whether this event requires a watch re-start, but
// not a full re-sync.
func eventRestartsWatch(e watch.Event) bool

// Closes a specific watcher
func closeWatcher(k key) 

// Closes all watchers (iterates over map and calls closeWatcher)
func closeAllWatchers()
```